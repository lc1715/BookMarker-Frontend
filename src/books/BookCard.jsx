import { Link } from "react-router-dom";
import removeHtmlTags from '../common/removeTags';
import defaultBookImage from '../assets/bookImage.jpg'
//Material UI
import Box from '@mui/material/Box';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';


/** BookCard 
 * 
 * - receives book as prop
 * - check if book contains isbn or volumeId property 
 * - {book} = {isbn or volumeId, title, author, description, image}
 * - show the book info 
 * 
 * - component using BookCard: BookList
 */

function BookCard({ book }) {
    console.log('book=', book)

    let bookId;
    let bookIdType;

    if (book.isbn) {
        bookId = book.isbn;
        bookIdType = 'isbn';
    } else {
        bookId = book.volumeId;
        bookIdType = 'volumeId'
    }
    console.log('book.descriptiopn', book.description)
    console.log('book.image', book.image)

    return (
        <Card sx={{ minWidth: { xs: 275, lg: 265 }, maxWidth: 150, minHeight: { xs: 500, lg: 485 }, mx: 5.5, mb: 5, backgroundColor: '#FCFBF4' }}>
            < Link to={`/books/${bookIdType}/${bookId}`} style={{ textDecoration: 'none' }}>
                <Box component="section" sx={{ p: 9, pt: 2, pb: 3 }}>
                    <CardMedia
                        component="img"
                        alt="book image"
                        height="auto"
                        image={book.image ? book.image : defaultBookImage}
                        sx={{ borderRadius: 2, width: { xs: '100%', lg: '99%' } }}
                    />
                </Box>

                <CardContent>
                    <Typography sx={{ color: 'text.primary', fontSize: { xs: 19, lg: 18 }, fontWeight: 'bold', textAlign: 'center' }}>
                        {book.title}
                    </Typography>

                    <Typography sx={{ color: 'text.primary', fontSize: { xs: 17, lg: 16 }, mb: 1.5, mt: 1.5, textAlign: 'center' }} >
                        By {book.author}
                    </Typography>

                    <Typography sx={{ color: 'text.primary', fontSize: { xs: 17, lg: 16 }, mb: 3, mt: 1.5, mx: 2, textAlign: 'center' }}>
                        {!book.description ? '' : removeHtmlTags(book.description).length > 200 ? removeHtmlTags(book.description).slice(0, 200) + '...' : removeHtmlTags(book.description)}
                    </Typography>
                </CardContent>
            </Link >
        </Card >
    )
};

export default BookCard;
