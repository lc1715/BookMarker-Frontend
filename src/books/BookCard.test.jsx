import { it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import BookCard from './BookCard';

const book = {
    title: 'title1',
    author: 'auth1',
    volumeId: 'vol1',
    category: 'cat1',
    description: 'desc1',
    image: 'img1',
    publisher: 'pub1',
}

it('renders the BookCard component', () => {
    render(
        <MemoryRouter>
            <BookCard book={book} />
        </MemoryRouter>
    )
})

it('matches with snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <BookCard book={book} />
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
})

it('renders the book information', () => {
    render(
        <MemoryRouter>
            <BookCard book={book} />
        </MemoryRouter>
    )

    expect(screen.getByText('title1')).toBeInTheDocument();
})
