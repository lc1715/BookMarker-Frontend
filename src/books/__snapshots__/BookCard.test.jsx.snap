// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches with snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ek9gvc-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <a
      href="/books/volumeId/vol1"
      style="text-decoration: none;"
    >
      <section
        class="MuiBox-root css-pzumkx"
      >
        <img
          alt="book image"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-qb6ft1-MuiCardMedia-root"
          height="auto"
          src="img1"
        />
      </section>
      <div
        class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1oth169-MuiTypography-root"
        >
          title1
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-l5k6vd-MuiTypography-root"
        >
          By auth1
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1a5af2-MuiTypography-root"
        >
          desc1
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
